### ISSUE TITLE
Request for LAPACK, XNNPPACK support for mobile pytorch model of react-native app

### ISSUE THREAD

**Issue by nh9k** [pytorch/pytorch#84109]
### 🚀 The feature, motivation and pitch

Hello!  Many thanks for contributions.
I have problems while developing my react-native app.

Could pytorch team support both LAPACK and XNNPACK? 

**1) Request for LAPACK support**

My model can't load at `react-native` app using `react-native-pytorch-core`
Related github is here: https://github.com/facebookresearch/playtorch

**React-native app code(.js):**
```
import {
  media,
  MobileModel,
  torch,
  torchvision,
} from 'react-native-pytorch-core';
...
console.log('Loading model: ', MODEL_URL);
const filePath = await MobileModel.download(require(MODEL_URL));
console.log('Local path of model is: ', filePath);
model = await torch.jit._loadForMobile(filePath);
console.log('Model successfully loaded');
```

**React-native app error(node.js):**
```
{"message": "Calling torch.linalg.lstsq on a CPU tensor requires compiling PyTorch with LAPACK. Please use PyTorch built with LAPACK support.

  Debug info for handle(s): debug_handles:{-1}, was not found.

Exception raised from apply_lstsq at ../aten/src/ATen/native/BatchLinearAlgebraKernel.cpp:559 (most recent call first):
(no backtrace available)"}
```

**My pytorch model code(.py):**
```
import torch, torchvision
from typing import List, Dict
import torchvision.transforms.functional as F

class WrapPerspectiveCrop(torch.nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, inputs: torch.Tensor, points: List[List[int]]):
        size_points = [[0,0], [inputs.shape[2],0] , [inputs.shape[2],inputs.shape[1]], [0,inputs.shape[1]]]
        inputs = F.perspective(inputs, points, size_points)
        return inputs

    
crop = WrapPerspectiveCrop()
scripted_model = torch.jit.script(crop)
scripted_model.save("wrap_perspective.pt")


import torch
from torch.utils.mobile_optimizer import optimize_for_mobile

optimized_scripted_module=optimize_for_mobile(scripted_model)
optimized_scripted_module._save_for_lite_interpreter("wrap_perspective.ptl")
```

My model is successful at both load and inference python code. 

Here is my related issue! 
https://github.com/facebookresearch/playtorch/issues/114

<br>

**2) Request for XNNPACK support**

My another model can't inference image tensor at `react-native` app using `react-native-pytorch-core`

**React-native app code(.js):**
```
 // Get image width and height
  const imageWidth = image.getWidth();
  const imageHeight = image.getHeight();

  // Convert image to blob, which is a byte representation of the image
  // in the format height (H), width (W), and channels (C), or HWC for short
  const blob = media.toBlob(image);

  // Get a tensor from image the blob and also define in what format
  // the image blob is.
  let tensor = torch.fromBlob(blob, [imageHeight, imageWidth, 3]);

  // Rearrange the tensor shape to be [CHW]
  tensor = tensor.permute([2, 0, 1]);

  // Unsqueeze adds 1 leading dimension to the tensor
  const formattedInputTensor = tensor.unsqueeze(0);

   // Run inference
  const output = await model.forward(formattedInputTensor);
```

**React-native app error(node.js):**
```
{"message": "XNNPACK Convolution not usable! Reason: The provided input tensor is either invalid or unsupported by XNNPACK.

  Debug info for handle(s): debug_handles:{-1}, was not found.

Exception raised from run at ../aten/src/ATen/native/xnnpack/Convolution.cpp:297 (most recent call first):
(no backtrace available)"}
```

My model is successful at both load and inference python code. 

**My pytorch model inference code(.py):**
```
import cv2
import torch , torchvision
import numpy as np


model = torch.jit.load('my_model.ptl')

img = cv2.imread("sample.jpg")
img_tensor = torch.as_tensor(np.expand_dims(img.astype("float32").transpose(2, 0, 1) , axis = 0))
outputs = model(img_tensor);

print(outputs)
```

If anyone has an alternative solution, please let me know any leads.
Thank you!

**Comment by raziel** [comment]
Hi @nh9k , this is not a core pytorch issue so you should log the issue with playtorch here: https://github.com/facebookresearch/playtorch/issues/new?assignees=&labels=Use%20Case&template=use_case.yml

**Comment by raziel** [comment]
Ah I see you already logged an issue there: https://github.com/facebookresearch/playtorch/issues/114

To answer your questions, we don't have any plans to add LAPACK support. Every such backend comes with non-trivial work to figure out and we don't see any requests to use lapack nor clear benefit from it.

Regarding XNNPack, we do have some support and plan to extend support further. @kimishpatel do you know what's the issue here?

Also, in general @nh9k , when you file bugs against pytorch it would be best if you give us pytorch code (i.e. python) as we don't work with the JS native code.

**Comment by kimishpatel** [comment]
@nh9k, do you know if you build pytorch with XNNPACK. Or whether playtorch is build with XNNPACK. There is USE_XNNPACK flag while building pytorch for OSS.

Regarding LAPACK support in pytorch. This has been extensively discussed in the past. Unfortunately we haven't really had to time add this on the mobile side. My apologies but cannot provide any timelines for this atm either.

**Comment by raziel** [comment]
Thanks @kimishpatel 

I think then the discussion should continue in the playtorch ticket https://github.com/facebookresearch/playtorch/issues/114
I'll close the issue (we took note of the LAPACK request), but we can reopen if there's really something missing on the pytorch side.

Thanks @nh9k for the feedback.

**Comment by nh9k** [comment]
Thank @raziel, @kimishpatel!

I do not understand fully about XNNPACK and LAPACK.
Thank you for discuss about this problems.

I think I will use opencv for react-native as the final alternative to LAPACK. 
If LAPACAK is added, we can use function like `torchvision.transforms.functional.perspective` at mobile.
Then i think, we can utilize more torchvision function with pytorch.

Regarding XNNPACK, i think it will solve this problem if i quantize my model to INT type.
There is was no attempt for building XNNPACK from pytorch or playtorch.

Thanks @raziel, @kimishpatel for the reply.
